### ISSUE TITLE
Some guidance on MsBuild Integration

### ISSUE THREAD

**Issue by sjdirect** [mkaring/ConfuserEx#110]
This is a GREAT project, thanks so much!!!!

I've installed the Confuser.MSBuild nuget package (2.0.0-alpha-gd91742f4c6) into the project that I want obfuscated. I don't see any difference in the output dlls/nugetpackage. No build output either. 

Are there any more docs or examples for the MSBuild task integration? I assumed that installation of the nuget package to the project you want obfuscated.

**Comment by sjdirect** [comment]
I see the msbuild target files but for some reason when installed on my .net standard 2.0 project the only change is the addition of the nuget package reference, no other changes to any source files.

**Comment by mkaring** [comment]
The MSBuild integration is not really documented yet. Sorry for that. There are a few things you need to do to get it working.

1. You need a `*.crproj` file that contains the configuration for the obfuscation. This is required to set what configurations are used. You can set this file up with the ConfuserEx UI and without any target assembly. Just setup the rules globally. There main assembly and the references are added automatically. If you don't add this file all configurations will be assumed with there default values. In this case you have to use the [`ObfuscationAttribute`][1] to set the configuration. The default name for that file is the same as the project file just with the `.crproj` extension. You can alternatively set the `ConfuserProject` Property in the MSBuild file to change the name
2. You need to set the `Obfuscate` property in the MSBuild project to `true`.

The output is controlled by two properties in the MSBuild project:
1. `ConfuserOutDir` contains the output directory of the obfuscated assemblies. It defaults to the normal output directory with an additional sub-directory named "confused".
2. `ConfuserReplaceOutput` contains a boolean value that defaults to `false`. In case this option is enabled, `ConfuserOutDir` has no effect anymore and instead the default output assembly is replaced.

I don't recommend to use the second one in case you are creating your debug builds, because this option will disrupt your debugging experience. 

[1]: https://docs.microsoft.com/dotnet/api/system.reflection.obfuscationattribute?view=netframework-4.8

**Comment by sjdirect** [comment]
I really appreciate the help (sorry i got tied up the last few days). I've tried your instructions (or at least my interpretation of them) and I can't get it to work.  I think a working project might help me better understand your instructions. I created a simple project on github...

TestApp
  TestApp (.net core console app that references TestApp.Lib)
  TestApp.Lib (.net standard class library which generates a nuget package on build, THIS ONE NEEDS TO BE OBFUSCATED ON BUILD BEFORE THE NUGET PACKAGE IS GENERATED)

Could you possibly create a pull request on this [project repo](https://github.com/sjdirect/TestApp) as an example to obfuscate the TestApp.Lib.dll with some basic protections on build as you described above? I also assume that any setup would run the obfuscate task on the TestApp.Lib.dll file before the nuget package is created by visual studio? Also assume that a build server would work as well. Are these assumptions correct?

Thanks Again!!!

**Comment by sjdirect** [comment]
Hi @mkaring any chance you can update that repo above as a working example of the MSBuild integration?

**Comment by mkaring** [comment]
I'll get on that in the next couple of days.

**Comment by mkaring** [comment]
Okay so there is a bug in there. The task creates the output in a separate directory without any issues, but if you want it to replace the real output of the compilation and have it packed into the Nuget package automatically, it does not work.

I'll schedule the fix for this in the next release for the 1.x branch of ConfuserEx. For now I'll use the build from the buildserver to update your reference repository.

**Comment by sjdirect** [comment]
Appreciate your time on this.

**Comment by ThomasBarnekow** [comment]
I tried to obfuscate a project that targets net472, using the NuGet package version 2.0.0-alpha-0191 from nuget.org and the guidance included in the PR. This leads to a critical error when building the project.

Here are the properties, noting that the error occurs with and without the `ConfuserReplaceOutput` property.
```
  <PropertyGroup>
    <Obfuscate>true</Obfuscate>
    <ConfuserReplaceOutput>true</ConfuserReplaceOutput>
  </PropertyGroup>
```

Here's the error message, which disappears if the above property group is removed.
```
...
1>[CRITICAL] Unknown error occurred.
...
```

**Comment by mkaring** [comment]
@ThomasBarnekow: Please use the released version 1.3.1. The MSBuild integration of this version works, the 2.0 version is not ready and the MSBuild integration isn't working right now.

**Comment by ThomasBarnekow** [comment]
@mkaring, thanks for the very quick reply. Will try that right now.

**Comment by ThomasBarnekow** [comment]
@mkaring, I have tried this with version 1.3.1 but it does not do anything (i.e., the assembly is not obfuscated and there is no obfuscation-related build output whatsoever). Maybe I am not configuring it correctly.

I have the following property group for a separate Confuse configuration, in which I added the `Optimize` and `ConfuserReplaceOutput` properties to existing properties used with ConfuserEx before.

```xml
  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Confuse|AnyCPU'">
    <Optimize>true</Optimize>
    <Obfuscate>true</Obfuscate> <!-- Added -->
    <ConfuserReplaceOutput>true</ConfuserReplaceOutput> <!-- Added -->
    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
    <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
  </PropertyGroup>
```

This is the `.crproj` file, which I have been using with ConfuserEx for quite some time:

```xml
<project outputDir="." baseDir="bin\Confuse" xmlns="http://confuser.codeplex.com">
  <rule preset="none" pattern="true">
    <protection id="anti debug" />
    <protection id="anti ildasm" />
    <protection id="ctrl flow" />
    <protection id="rename" />
    <protection id="ref proxy" />
  </rule>
  <module path="..." />
</project>
```

**Comment by ThomasBarnekow** [comment]
@mkaring, I have played around some more. When I remove the `ConfuserReplaceOutput` property, it works and the confused DLL is created in a `confused` subfolder of my `bin\Confuse` output folder. 

When building on the command line with `ConfuserReplaceOutput` set to `true`, there is build output (see below), but `MyAssembly.dll` is not obfuscated. Not sure whether the obfuscated DLL is written somewhere else. I've tried it with `baseDir="bin\Confuse"` and `baseDir="."`.

```
ConfuseAssembly:
  Confuser.Core 1.3.1+cf4240c129 Copyright © 2014 Ki, 2018 Martin Karing
  Running on Microsoft Windows NT 10.0.18362.0, .NET Framework v4.0.30319.42000, 32 bits
  Discovered 10 protections, 1 packers.
  Loading input modules...
  Loading 'MyAssembly.dll'...
  Initializing...
  Resolving dependencies...
  Processing module 'MyAssembly.dll'...
  Writing module 'MyAssembly.dll'...
  Finalizing...
  Done.
```

Not being able to use the `ConfuserReplaceOutput` property would require undesired changes to my build process, which involves a larger number of projects. Thus, I'd really like to replace the Release DLL built in the first step to be replaced with the confused one

**Comment by mkaring** [comment]
Yes. As I wrote above, there is a bug in the MSBuild task that breaks this properly. I fixed the problem in the latest preview build for the 1.4 version.

You find a properly setup project here: https://github.com/mkaring/TestApp
Make sure that you add the exact same version for the ConfuserEx `ProjectReference` and use the additional nuget source that is in the `NuGet.config`.

This will fix the `ConfuserReplaceOutput`. Once version 1.4 is being released, it will work in the regular release.

**Comment by ThomasBarnekow** [comment]
@mkaring, thanks. Using the 1.4 preview worked.

**Comment by sjdirect** [comment]
@mkaring using the 1.4 preview worked for me as well. Appreciate you forking and using the TestApp as an example. No way I would have figured out the necessary changes without it.

**Comment by mkaring** [comment]
I am closing this issue. My final tests with the changes have concluded.
For another project it's required that the MSBuild integration of ConfuserEx is compatible with @Fody and that is still the case.

The patch will be released with the 1.4 Update.

**Comment by MateusMac** [comment]
Hey guys, I'm trying to use Nuget as well, I already created the .crproj and configured my .csproj. But when I try to compile I'm get the error

` Confuser.MSBuild.targets(66, 5): Cannot find protection with ID 'anti debug'.`

Here is my .crproj

`<project outputDir=".\Confused" baseDir=".\bin\Release\netstandard2.1\publish" xmlns="http://confuser.codeplex.com">
  <rule pattern="true" preset="normal" inherit="false">
    <protection id="anti debug" />
    <protection id="rename" />
    <protection id="resources" />
    <protection id="ref proxy" />
    <protection id="anti tamper" />
  </rule>
  <module path="My.dll" />
</project>`

And my .csproj

` <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
        <Optimize>true</Optimize>
        <Obfuscate>true</Obfuscate>
        <ConfuserReplaceOutput>true</ConfuserReplaceOutput>
        <AppendTargetFrameworkToOutputPath>true</AppendTargetFrameworkToOutputPath>
        <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
    </PropertyGroup>`

Any ideias?
